kolibri/core/assets/src/api_resource.js

```diff
@@ -0,0 +1,263 @@
+   * @param {String} id - The primary key of the Model instance.
+   * @returns {Model} - Returns a Model instance.
+   */
+  getModel(id) {
+    let model;
+    if (!this.models[id]) {
+      model = new Model({ [this.idKey]: id }, this);
+    } else {
+      model = this.models[id];
+    }
+    return model;
+  }
+
+  /**
+   * Add a model to the resource for deduplication, dirty checking, and tracking purposes.
+   * @param {Object|Model} model - Either the data for the model to add, or the Model itself.
+   * @returns {Model} - Returns the instantiated Model.
+   */
+  addModel(model) {
+    if (!(model instanceof Model)) {
+      model = new Model(model, { resource: this }); // eslint-disable-line no-param-reassign
+    }
+    if (!this.models[model.id]) {
+      this.models[model.id] = model;
+    } else {
+      this.models[model.id].set(model.attributes);
+    }
+    return this.models[model.id];
+  }
+  get urls() {
+    return this.kolibri.urls;
+  }
+  get modelUrl() {
+    return this.urls[`${this.name}_detail`];
+  }
+  get collectionUrl() {
+    return this.urls[`${this.name}_list`];
+  }
+}
+
+/** Class to manage all API resources.
+ *  This is instantiated and attached to the core app constructor, and its methods exposed there.
+ *  This means that a particular Resource should only be instantiated once during the lifecycle
+ * of the app, allowing for easy caching, as all Model instances can be shared in the central
+ * resource.
+ */
+class ResourceManager {
+  /**
+   * Instantiate a Resource Manager to manage the creation of Resources.
+   * @param {Kolibri} kolibri - The current instantiated instance of the core app - needed to
+  * reference the urls.
+   */
+  constructor(kolibri) {
+    this.kolibri = kolibri;
+    this._resources = {};
+  }
+
+  /**
+   * Register a resource with the resource manager. Only one resource of a particular name can be
+   * registered.
+   * @param {String} name - The Django REST framework name for the API endpoint.
+   * @param {String} idKey - The primary key field for the API endpoint, defaults to 'id'.
+   * @param {Class} [ResourceClass=Resource] - The subclass of Resource to use in registering the
+   * resource. This can be used to register a resource with specific subclassed behaviour for that
+   * resource.
+   * @returns {Resource} - Return the instantiated Resource.
+   */
+  registerResource(name, idKey = 'id', ResourceClass = Resource) {
+    if (name && !this._resources[name]) {
+      this._resources[name] = new ResourceClass({ name, idKey, kolibri: this.kolibri });
+    } else {
+      if (!name) {
+        throw new TypeError('A resource must have a defined resource name!');
+      } else {
+        throw new TypeError('A resource with that name has already been registered!');
+      }
+    }
+    return this._resources[name];
+  }
+
+  /**
+   * Get a resource from the resource manager. For convenience, will register a resource for that
+   * name if one does not exist.
```

Comment:
Do we want to allow generic, untyped resources? It may be more consistent to explicitly define a resource definition for every API name used on the client-side. (I realize this made it easier to port the management plugin code)